barretenberg/cpp/src/barretenberg/eccvm/transcript_builder.hpp

```diff
@@ -104,225 +130,397 @@ class ECCVMTranscriptBuilder {
-            row.z1_zero = z1_zero;
-            row.z2_zero = z2_zero;
-            row.opcode = Opcode{ .add = entry.add, .mul = entry.mul, .eq = entry.eq, .reset = entry.reset }.value();
+            // update the accumulators
             accumulator_trace[i] = state.accumulator;
             msm_accumulator_trace[i] = msm_transition ? updated_state.msm_accumulator : Element::infinity();
             intermediate_accumulator_trace[i] =
                 msm_transition ? (updated_state.msm_accumulator - offset_generator()) : Element::infinity();
-            if (entry.mul && next_not_msm && !row.accumulator_empty) {
-                state.msm_accumulator = offset_generator();
-            }
 
             state = updated_state;
 
-            if (entry.mul && next_not_msm) {
+            if (is_mul && next_not_msm) {
                 state.msm_accumulator = offset_generator();
             }
         }
+        // Compute affine coordinates of accumulated points
+        normalize_accumulators(accumulator_trace, msm_accumulator_trace, intermediate_accumulator_trace);
+
+        // Add required affine coordinates to the transcript
+        add_affine_coordinates_to_transcript(
+            transcript_state, accumulator_trace, msm_accumulator_trace, intermediate_accumulator_trace);
+
+        // Process the slopes when adding points or results of MSMs
+        for (size_t i = 0; i < accumulator_trace.size(); ++i) {
+            TranscriptRow& row = transcript_state[i + 1];
+            const bool msm_transition = row.msm_transition;
+
+            const VMOperation& entry = vm_operations[i];
+            const bool is_add = entry.add;
+
+            if (msm_transition || is_add) {
+                // compute the differences between point coordinates
+                compute_inverse_trace_coordinates(msm_transition,
+                                                  row,
+                                                  intermediate_accumulator_trace[i],
+                                                  transcript_msm_x_inverse_trace[i],
+                                                  msm_accumulator_trace[i],
+                                                  accumulator_trace[i],
+                                                  inverse_trace_x[i],
+                                                  inverse_trace_y[i]);
+
+                // compute the numerators and denominators of slopes between the points
+                compute_lambda_numerator_and_denominator(row,
+                                                         entry,
+                                                         intermediate_accumulator_trace[i],
+                                                         accumulator_trace[i],
+                                                         add_lambda_numerator[i],
+                                                         add_lambda_denominator[i]);
+            } else {
+                row.transcript_add_x_equal = 0;
```

Comment:
what does this do?

Reply:
the slopes are only used when there's an msm transition or an addition